### PR TITLE
fix failure with devel after stricteffects enabled, fix #56

### LINKED ISSUE #56: Library errors with devel after stricteffects enabled
compiling the first readme example with current devel throws an error.

example file:
```nim
import toml_serialization

let rawToml = """
[server]
  name = "TOML Server"
  port = 8005
"""
var x = Toml.decode(rawToml, string, "server.name")
assert x == "TOML Server"
```

error reported:
```
C:\Users\ppeterlongo\.nimble\pkgs\toml_serialization-0.2.3\toml_serialization\types.nim(172, 25) template/generic instantiation of `==` from here
C:\Users\ppeterlongo\.choosenim\toolchains\nim-#devel\lib\system\comparisons.nim(309, 6) Error: '==' can have side effects
sideEffect` '=='
>> C:\Users\ppeterlongo\.nimble\pkgs\toml_serialization-0.2.3\toml_serialization\types.nim(144, 6) Hint: '==' called by '=='
```

see also https://github.com/nim-lang/Nim/pull/20697

it seems adding a noSideEffects annotation to `==` should fix this. I will check if that is indeed the case and come back with a PR.

### PR DESCRIPTION
**Pull request by pietroppeter**
see #56

### CODE REVIEW

**Comment by pietroppeter** [comment]
I am expecting error reported in #56 to show when CI runs. not sure how to proceed with fixing the new error

**Comment by jangko** [comment]
instead of using nosideeffect pragma, it will be cleaner and shorter to use `func` to replace the `proc`.

the unlisted exception indeed annoying because we are not sure where to look into for the source of exception.

**Comment by pietroppeter** [comment]
- changed annotated proc to func
- manually track exceptions as in #55 (following [a suggestion](https://github.com/nim-lang/Nim/pull/20697#issuecomment-1297898150) from @ringabout)
- I could add a protection by nim version (apply last change on tracking exceptions only for >= 1.7) but for the moment the simple example above compiles with both devel and stable (using choosenim locally)
- although locally nimble test fails with not a particularly revealing stack trace:
```
[NimScript] exec: nim cpp --threads:off -d:release --outdir:build -r --hints:off --skipParentCfg --styleCheck:usages --styleCheck:error tests/test_all
Error: execution of an external program failed: 'C:\Users\ppeterlongo\Documents\toml_serialization\build\test_all.exe'
stack trace: (most recent call last)
C:\Users\PPETER~1\AppData\Local\Temp\nimblecache-0\nimscriptapi_403350235.nim(187, 16)
C:\Users\ppeterlongo\Documents\toml_serialization\toml_serialization.nimble(49, 8) testTask
C:\Users\ppeterlongo\Documents\toml_serialization\toml_serialization.nimble(42, 8) test
C:\Users\ppeterlongo\.choosenim\toolchains\nim-#devel\lib\system\nimscript.nim(264, 7) exec
C:\Users\ppeterlongo\.choosenim\toolchains\nim-#devel\lib\system\nimscript.nim(264, 7) Error: unhandled exception: FAILED: nim cpp --threads:off -d:release --outdir:build -r --hints:off --skipParentCfg --styleCheck:usages 
--styleCheck:error tests/test_all [OSError]
     Error: Exception raised during nimble script execution
```

File: toml_serialization/lexer.nim [hunk 1]
```diff
@@ -1500,7 +1500,7 @@ proc parseNumOrDate*[T](lex: var TomlLexer, value: var T) =
       raiseIllegalChar(lex, next)
     break
 
-proc parseValue*[T](lex: var TomlLexer, value: var T)
+proc parseValue*[T](lex: var TomlLexer, value: var T) {.raises: [IOError, TomlError].}
```
Comment (arnetheduck) [inline toml_serialization/lexer.nim:1503]:
```suggestion
proc parseValue*[T](lex: var TomlLexer, value: var T) {.raises: [Defect, IOError, TomlError].}
```
for 1.2 compat

File: toml_serialization/types.nim [hunk 2]
```diff
@@ -126,9 +126,9 @@ when tomlOrderedTable:
       let value {.inject.} = addr(valx)
       body1
 
-proc `==`*(a, b: TomlValueRef): bool
+func `==`*(a, b: TomlValueRef): bool
```
Comment (arnetheduck) [inline toml_serialization/types.nim:129]:
```suggestion
func `==`*(a, b: TomlValueRef): bool
```

**Comment by arnetheduck** [comment]
> nimble test fails 

what about running `nim cpp --threads:off -d:release --outdir:build -r --hints:off --skipParentCfg --styleCheck:usages --styleCheck:error tests/test_all` manually?

**Comment by pietroppeter** [comment]
> > nimble test fails
> 
> what about running `nim cpp --threads:off -d:release --outdir:build -r --hints:off --skipParentCfg --styleCheck:usages --styleCheck:error tests/test_all` manually?

I did but it does not provides additional info

**Comment by pietroppeter** [comment]
ok, looking at some of the failing checks I can see that in some cases (e.g. windows-amd64-c-(Nim-version-1-6), but also linux and possibly others) an error is reported during compilation:
```
/home/runner/work/nim-toml-serialization/nim-toml-serialization/nim-toml-serialization/toml_serialization/private/array_reader.nim(40, 12) Error: 'readArray' is not GC-safe as it calls 'readValue'
```
in other cases (in particular at least some of those using devel, which include my local setup of windows-amd64), the error is not reported during compilation and running fails without a stacktrace.

**Comment by Menduist** [comment]
This should help: https://github.com/status-im/nim-serialization/pull/44

**Comment by jangko** [comment]
until issue in nim-serialization solved,  https://github.com/status-im/nim-serialization/issues/45, we cannot do much to remove the segfaults.

**Pull request merged**